## Overview

### I have been given a dataset with recruits personality scores and the department that they applied for. 
### I will clean the data, and find insights from the data.

### import required packages

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Read in the dataset personality_scores.csv.

In [2]:
scores = pd.read_csv('personality_scores.csv', delimiter =';')

### Examine the data frame for duplicates and NaN values.

In [11]:
scores.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ... Unnamed: 60  \
0                                            (5, 5)  ...         NaN   
1                                            (5, 5)  ...         NaN   
2                                            (5, 5)  ...         NaN   
3                                            (5, 3)  ...         NaN   
4                                            (5, 5)  ...         NaN   

  Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65 Unnamed: 66  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK  
0         NaN         NaN            NaN  
1         NaN         NaN            NaN  
2         NaN         NaN            NaN  
3         NaN         NaN            NaN  
4         NaN         NaN            NaN  

[5 rows x 70 columns]

In [10]:
scores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 70 columns):
ID                                                                          1555 non-null int64
Section 5 of 6 [I am always prepared.]                                      1555 non-null object
Section 5 of 6 [I am easily disturbed.]                                     1555 non-null object
Section 5 of 6 [I am exacting (demanding) in my work.]                      1555 non-null object
Section 5 of 6 [I am full of ideas.]                                        1555 non-null object
Section 5 of 6 [I am interested in people.]                                 1555 non-null object
Section 5 of 6 [I am not interested in abstract ideas.]                     1555 non-null object
Section 5 of 6 [I am not interested in other people's problems.]            1555 non-null object
Section 5 of 6 [I am not really interested in others.]                      1555 non-null object
Section 5 of 6 [I am 

In [9]:
scores.isna().sum().sum()

29545

### There aren't any duplicate columns but there are 29545 NaN values in the dataset. 
### I cannot use the columns with NaN values so I will drop them.

In [20]:
scores_clean = scores.dropna(axis = 1, how ='any')

In [21]:
scores_clean.isna().sum().sum()

0

In [22]:
scores_clean.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ...  \
0                                            (5, 5)  ...   
1                                            (5, 5)  ...   
2                                            (5, 5)  ...   
3                                            (5, 3)  ...   
4                                            (5, 5)  ...   

  Section 5 of 6 [I often forget to put things back in their proper place]  \
0                                             (3, 5)                         
1                                             (3, 5)                         
2                                             (3, 5)                         
3                                             (3, 1)                         
4                                             (3, 5)                         

  Section 5 of 6 [I pay attention to details.]  \
0                                       (3, 5)   
1                                       (3, 1)   
2                                       (3, 5)   
3                                       (3, 5)   
4                                       (3, 5)   

  Section 5 of 6 [I seldom feel blue (down).]  \
0                                      (4, 3)   
1                                      (4, 1)   
2                                      (4, 1)   
3                                      (4, 1)   
4        

### I have dropped all the NaN values and now I can work with the dataset
### I see that the index and ID columns are the same so I'm going to make the ID the index

In [23]:
scores_clean = scores_clean.set_index('ID')

In [24]:
scores_clean.head()

Section 5 of 6 [I am always prepared.]  \
ID                                          
0                                  (3, 5)   
1                                  (3, 5)   
2                                  (3, 5)   
3                                  (3, 5)   
4                                  (3, 3)   

   Section 5 of 6 [I am easily disturbed.]  \
ID                                           
0                                   (4, 5)   
1                                   (4, 5)   
2                                   (4, 3)   
3                                   (4, 5)   
4                                   (4, 5)   

   Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                          
0                                              (3, 5)       
1                                              (3, 5)       
2                                              (3, 3)       
3                                              (3, 3)       
4                                              (3, 3)       

   Section 5 of 6 [I am full of ideas.]  \
ID                                        
0                                (5, 5)   
1                                (5, 5)   
2                                (5, 5)   
3                                (5, 5)   
4                                (5, 3)   

   Section 5 of 6 [I am interested in people.]  \
ID                                               
0                                       (2, 3)   
1                                       (2, 5)   
2                                       (2, 5)   
3                                       (2, 5)   
4                                       (2, 3)   

   Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                           
0                                              (5, 3)        
1                                              (5, 3)        
2                                              (5, 5)        
3                                              (5, 3)        
4                                              (5, 3)        

   Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                    
0                                              (2, 3)                 
1                                              (2, 5)                 
2                                              (2, 5)                 
3                                              (2, 3)                 
4                                              (2, 3)                 

   Section 5 of 6 [I am not really interested in others.]  \
ID                                                          
0                                              (2, 5)       
1                                              (2, 5)       
2                                              (2, 5)       
3                                              (2, 3)       
4                                              (2, 3)       

   Section 5 of 6 [I am quick to understand things.]  \
ID                                                     
0                                             (5, 5)   
1                                             (5, 5)   
2                                             (5, 5)   
3                                             (5, 3)   
4                                             (5, 5)   

   Section 5 of 6 [I am quiet around strangers.]  ...  \
ID                                                ...   
0                                         (1, 3)  ...   
1                                         (1, 3)  ...   
2                                         (1, 1)  ...   
3                                         (1, 3)  ...   
4                                         (1, 1)  ...   

   Section 5 of 6 [I often forget to put things back in their proper place]  \
ID                                                

### Write functions that will (1) convert the answer on each of the personality test questions to a number, as set out in scoring.md, and (2) calculate total scores for each of the subscales (conscientiousness, emotional stability, openness to new experience, agreeableness, extraversion).

In [25]:
def sum_per_subscale(scores_per_person, subscale):
    subscale_sum = sum([eval(scores_per_person[i])[1] for i in range(50) if eval(scores_per_person[i])[0] == subscale])
    return(subscale_sum)

sum_per_subscale(scores_clean.loc[0], 1)

scores_clean['Extraversion'] = scores_clean.apply(lambda row: sum_per_subscale(row, 1), axis=1)
scores_clean['Agreeableness'] = scores_clean.apply(lambda row: sum_per_subscale(row, 2), axis=1)
scores_clean['Conscientiousness'] = scores_clean.apply(lambda row: sum_per_subscale(row, 3), axis=1)
scores_clean['EmotionalStability'] = scores_clean.apply(lambda row: sum_per_subscale(row, 4), axis=1)
scores_clean['Intellect'] = scores_clean.apply(lambda row: sum_per_subscale(row, 5), axis=1)

### The function that I used above was worked on by nadja our mentor and us c13. 

In [26]:
scores_clean.head()

Section 5 of 6 [I am always prepared.]  \
ID                                          
0                                  (3, 5)   
1                                  (3, 5)   
2                                  (3, 5)   
3                                  (3, 5)   
4                                  (3, 3)   

   Section 5 of 6 [I am easily disturbed.]  \
ID                                           
0                                   (4, 5)   
1                                   (4, 5)   
2                                   (4, 3)   
3                                   (4, 5)   
4                                   (4, 5)   

   Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                          
0                                              (3, 5)       
1                                              (3, 5)       
2                                              (3, 3)       
3                                              (3, 3)       
4                                              (3, 3)       

   Section 5 of 6 [I am full of ideas.]  \
ID                                        
0                                (5, 5)   
1                                (5, 5)   
2                                (5, 5)   
3                                (5, 5)   
4                                (5, 3)   

   Section 5 of 6 [I am interested in people.]  \
ID                                               
0                                       (2, 3)   
1                                       (2, 5)   
2                                       (2, 5)   
3                                       (2, 5)   
4                                       (2, 3)   

   Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                           
0                                              (5, 3)        
1                                              (5, 3)        
2                                              (5, 5)        
3                                              (5, 3)        
4                                              (5, 3)        

   Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                    
0                                              (2, 3)                 
1                                              (2, 5)                 
2                                              (2, 5)                 
3                                              (2, 3)                 
4                                              (2, 3)                 

   Section 5 of 6 [I am not really interested in others.]  \
ID                                                          
0                                              (2, 5)       
1                                              (2, 5)       
2                                              (2, 5)       
3                                              (2, 3)       
4                                              (2, 3)       

   Section 5 of 6 [I am quick to understand things.]  \
ID                                                     
0                                             (5, 5)   
1                                             (5, 5)   
2                                             (5, 5)   
3                                             (5, 3)   
4                                             (5, 5)   

   Section 5 of 6 [I am quiet around strangers.]  ...  \
ID                                                ...   
0                                         (1, 3)  ...   
1                                         (1, 3)  ...   
2                                         (1, 1)  ...   
3                                         (1, 3)  ...   
4                                         (1, 1)  ...   

   Section 5 of 6 [I sympathize with others' feelings.]  \
ID                                                        
0          

### Create a function/functions that takes in subscale total, and for each subscale, creates a new column called "{subscale_name}_interpretation" (e.g., conscientiousness_interpretation) and applies the labels "low", "medium", or "high" according to a person's score on that subscale. Cut-off scores corresponding to the low, medium and high groups can be found in scoring.md.

In [32]:
Extraversion_low = scores_clean['Extraversion'].mean() - scores_clean['Extraversion'].std()/2
Extraversion_high = scores_clean['Extraversion'].mean() + scores_clean['Extraversion'].std()/2

Agreeableness_low = scores_clean['Agreeableness'].mean() - scores_clean['Agreeableness'].std()/2
Agreeableness_high = scores_clean['Agreeableness'].mean() + scores_clean['Agreeableness'].std()/2

Conscientiousness_low = scores_clean['Conscientiousness'].mean() - scores_clean['Conscientiousness'].std()/2
Conscientiousness_high = scores_clean['Conscientiousness'].mean() + scores_clean['Conscientiousness'].std()/2

Emotional_Stability_low = scores_clean['EmotionalStability'].mean() - scores_clean['EmotionalStability'].std()/2
Emotional_Stability_high = scores_clean['EmotionalStability'].mean() + scores_clean['EmotionalStability'].std()/2

Intellect_low = scores_clean['Intellect'].mean() + scores_clean['Intellect'].std()/2
Intellect_high = scores_clean['Intellect'].mean() + scores_clean['Intellect'].std()/2

### Extraversion Low, Mediam and High

In [57]:
Extraversion_interpretation = []

for row in scores_clean['Extraversion']:
    if row < Extraversion_low:
        Extraversion_interpretation.append('Low')
    elif row > Extraversion_high:
        Extraversion_interpretation.append('High')
    else:
        Extraversion_interpretation.append('Medium')
        
scores_clean['Extraversion_interpretation'] = Extraversion_interpretation

### Conscientiousness Low, Mediam and high

In [56]:
Conscientiousness_interpretation = []

for row in scores_clean['Conscientiousness']:
    if row < Conscientiousness_low:
        Conscientiousness_interpretation.append('Low')
    elif row > Conscientiousness_high:
        Conscientiousness_interpretation.append('High')
    else:
        Conscientiousness_interpretation.append('Medium')
        
scores_clean['Conscientiousness_interpretation'] = Conscientiousness_interpretation

### Emotional_Stability Low, Mediam and high

In [55]:
Emotional_Stability_interpretation = []

for row in scores_clean['EmotionalStability']:
    if row < Emotional_Stability_low:
        Emotional_Stability_interpretation.append('Low')
    elif row > Emotional_Stability_high:
        Emotional_Stability_interpretation.append('High')
    else:
        Emotional_Stability_interpretation.append('Medium')
        
scores_clean['Emotional_Stability_interpretation'] = Emotional_Stability_interpretation

### Intellect Low, Medium and High

In [54]:
Intellect_interpretation = []

for row in scores_clean['Intellect']:
    if row < Intellect_low:
        Intellect_interpretation.append('Low')
    elif row > Intellect_high:
        Intellect_interpretation.append('High')
    else:
        Intellect_interpretation.append('Medium')
        
scores_clean['Intellect_interpretation'] = Intellect_interpretation

In [59]:
scores_clean.head()

Section 5 of 6 [I am always prepared.]  \
ID                                          
0                                  (3, 5)   
1                                  (3, 5)   
2                                  (3, 5)   
3                                  (3, 5)   
4                                  (3, 3)   

   Section 5 of 6 [I am easily disturbed.]  \
ID                                           
0                                   (4, 5)   
1                                   (4, 5)   
2                                   (4, 3)   
3                                   (4, 5)   
4                                   (4, 5)   

   Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                          
0                                              (3, 5)       
1                                              (3, 5)       
2                                              (3, 3)       
3                                              (3, 3)       
4                                              (3, 3)       

   Section 5 of 6 [I am full of ideas.]  \
ID                                        
0                                (5, 5)   
1                                (5, 5)   
2                                (5, 5)   
3                                (5, 5)   
4                                (5, 3)   

   Section 5 of 6 [I am interested in people.]  \
ID                                               
0                                       (2, 3)   
1                                       (2, 5)   
2                                       (2, 5)   
3                                       (2, 5)   
4                                       (2, 3)   

   Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                           
0                                              (5, 3)        
1                                              (5, 3)        
2                                              (5, 5)        
3                                              (5, 3)        
4                                              (5, 3)        

   Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                    
0                                              (2, 3)                 
1                                              (2, 5)                 
2                                              (2, 5)                 
3                                              (2, 3)                 
4                                              (2, 3)                 

   Section 5 of 6 [I am not really interested in others.]  \
ID                                                          
0                                              (2, 5)       
1                                              (2, 5)       
2                                              (2, 5)       
3                                              (2, 3)       
4                                              (2, 3)       

   Section 5 of 6 [I am quick to understand things.]  \
ID                                                     
0                                             (5, 5)   
1                                             (5, 5)   
2                                             (5, 5)   
3                                             (5, 3)   
4                                             (5, 5)   

   Section 5 of 6 [I am quiet around strangers.]  ...  \
ID                                                ...   
0                                         (1, 3)  ...   
1                                         (1, 3)  ...   
2                                         (1, 1)  ...   
3                                         (1, 3)  ...   
4                                         (1, 1)  ...   

   Section 5 of 6 [I worry about things.] Extraversion Agreeableness  \
ID                                                       

### Read in the data in department_infromation.csv and merge this data frame with the personality score data frame, keeping all applicants within the department data frame. Assert that the newly created merged data frame has the same amount of rows as the department data frame, and the expected number of columns.

In [60]:
departments = pd.read_csv('departments.csv', delimiter = ';')

In [61]:
departments.head()

ID Department  Unnamed: 2  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0   0       Data         NaN         NaN         NaN         NaN         NaN   
1   1       Data         NaN         NaN         NaN         NaN         NaN   
2   2       Data         NaN         NaN         NaN         NaN         NaN   
3   3       Data         NaN         NaN         NaN         NaN         NaN   
4   4       Data         NaN         NaN         NaN         NaN         NaN   

   Unnamed: 7  Unnamed: 8  Unnamed: 9  ...  Unnamed: 11  Unnamed: 12  \
0         NaN         NaN         NaN  ...          NaN          NaN   
1         NaN         NaN         NaN  ...          NaN          NaN   
2         NaN         NaN         NaN  ...          NaN          NaN   
3         NaN         NaN         NaN  ...          NaN          NaN   
4         NaN         NaN         NaN  ...          NaN          NaN   

   Unnamed: 13  Unnamed: 14  Unnamed: 15  Unnamed: 16  Unnamed: 17  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 18  Unnamed: 19  IPIP_HIGH_RISK  
0          NaN          NaN             NaN  
1          NaN          NaN             NaN  
2          NaN          NaN             NaN  
3          NaN          NaN             NaN  
4          NaN          NaN             NaN  

[5 rows x 21 columns]

### i want to find out how many NaN values there are

In [62]:
departments.isnull().sum().sum()

29545

### I have noticed that there are a number of rows with Null values. I am going to use .dropna(axis = 1, how ='any') to drop these columns because I cannot use them

In [63]:
departments = departments.dropna(axis = 1, how ='any')

In [64]:
departments.head()

ID Department
0   0       Data
1   1       Data
2   2       Data
3   3       Data
4   4       Data

### I am going to make the index of the dataframe ID

In [65]:
departments = departments.set_index('ID')

In [66]:
departments.head()

Department
ID           
0        Data
1        Data
2        Data
3        Data
4        Data

### I will now join the two data sets

In [67]:
scores_join = scores_clean.join(departments)

In [68]:
scores_join.head()

Section 5 of 6 [I am always prepared.]  \
ID                                          
0                                  (3, 5)   
1                                  (3, 5)   
2                                  (3, 5)   
3                                  (3, 5)   
4                                  (3, 3)   

   Section 5 of 6 [I am easily disturbed.]  \
ID                                           
0                                   (4, 5)   
1                                   (4, 5)   
2                                   (4, 3)   
3                                   (4, 5)   
4                                   (4, 5)   

   Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                          
0                                              (3, 5)       
1                                              (3, 5)       
2                                              (3, 3)       
3                                              (3, 3)       
4                                              (3, 3)       

   Section 5 of 6 [I am full of ideas.]  \
ID                                        
0                                (5, 5)   
1                                (5, 5)   
2                                (5, 5)   
3                                (5, 5)   
4                                (5, 3)   

   Section 5 of 6 [I am interested in people.]  \
ID                                               
0                                       (2, 3)   
1                                       (2, 5)   
2                                       (2, 5)   
3                                       (2, 5)   
4                                       (2, 3)   

   Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                           
0                                              (5, 3)        
1                                              (5, 3)        
2                                              (5, 5)        
3                                              (5, 3)        
4                                              (5, 3)        

   Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                    
0                                              (2, 3)                 
1                                              (2, 5)                 
2                                              (2, 5)                 
3                                              (2, 3)                 
4                                              (2, 3)                 

   Section 5 of 6 [I am not really interested in others.]  \
ID                                                          
0                                              (2, 5)       
1                                              (2, 5)       
2                                              (2, 5)       
3                                              (2, 3)       
4                                              (2, 3)       

   Section 5 of 6 [I am quick to understand things.]  \
ID                                                     
0                                             (5, 5)   
1                                             (5, 5)   
2                                             (5, 5)   
3                                             (5, 3)   
4                                             (5, 5)   

   Section 5 of 6 [I am quiet around strangers.]  ... Extraversion  \
ID                                                ...                
0                                         (1, 3)  ...           30   
1                                         (1, 3)  ...           42   
2                                         (1, 1)  ...           28   
3                                         (1, 3)  ...           30   
4                                         (1, 1)  ...           28   

   Agreeableness Conscientiousness Emo